### ISSUE TITLE
Why don't Memory/ReadOnlyMemory have indexers?

### ISSUE THREAD

**Issue by shravan2x** [dotnet/runtime#105286]
[This Microsoft article](https://learn.microsoft.com/en-us/archive/msdn-magazine/2018/january/csharp-all-about-span-exploring-a-new-net-mainstay) hints an implementation like the one below for Memory/ReadOnlyMemory types in C#.

```cs
public readonly struct Memory<T>
{
  private readonly object _object;
  private readonly int _index;
  private readonly int _length;
  ...
}
```

Intuitively, this would suggest that accessing an element by index would have the same performance as accessing an array element by index. But @benaadams mentioned [here](https://github.com/dotnet/runtime/issues/11018#issuecomment-417642419) that adding an indexer to Memory has a severe performance cost associated. Could you explain why?

It seems I'm [not the only one](https://stackoverflow.com/questions/65143152/why-does-memoryt-not-have-an-indexer-while-spant-does) with this question.

**Comment by Wraith2** [comment]
My understanding of Memory and ReadOnlyMemory from listening to api reviews is that they exist only to let you pass spans to methods which do not allow ref structs, for example async methods. If you think of them as nothing more than a heap allocatable source of spans then the small surface area and lack of interfaces makes more sense.

**Comment by AlgorithmsAreCool** [comment]
TL;DR: It is an intentional design tradeoff that prioritized `[ReadOnly]Memory<T>` being cheap to create over being cheap to access.

Span<T> is just a wrapper around a pointer. Super lightweight. But `Memory<T>` is more complicated, and it can be backed by multiple types of memory including `strings`, `array`, and a fully customizable `MemoryManager<T>`. Each of these 3 types require their own handling path. Deciding which path to use and then using it has some overhead. We can see [how complex the getter is here.](https://github.com/dotnet/runtime/blob/cb7817fbfd4c582594197a762dacbde773dd92f3/src/libraries/System.Private.CoreLib/src/System/Memory.cs#L266)

If `Memory<T>` were a class, they could have possibly made this faster by using inheritance and virtual dispatch to get the correct path each time. (Think `StringMemory`, `ArrayMemory<T>` and `MemoryManagerMemory<T>`) but then we would have to pay increased for garbage collector pressure on a type that needs to be cheap to construct. 

The design that was chosen places this overhead into the `Span` property getter on the `Memory<T>` object, and you pay for it every time you access that property. But it isn't that much and in much better overall than needing to create tons of heap objects to tie up the GC. 

So back to your actual question, why not implement the accessor directly on `Memory<T>`? Well, because of the reason explained above, accessing the underlying span is rather expensive and people most commonly access the items repeatedly in a loop. Exposing an inefficient indexer would lead people into an obvious pit of failure of repeatedly paying the access overhead every step of the iteration. To force people onto the happy path, they chose not to define the indexer on Memory<T>.

**Comment by shravan2x** [comment]
Appreciate the answers! I'll go ahead and close.
